### ISSUE TITLE
Understanding WorkFlow for Training and Testing

### ISSUE THREAD

**Issue by tarunjainsagar** [facebookarchive/caffe2#352]
This is in continuation to #337
So now, my question is:

Ques. 01
I want to create a training network with FC, Sigmoid, SoftmaxWithLoss Layers.
Do I feed all the training data blob into workspace before running CreateNet for Training Model Definition ?

Ques. 02
If above statement is true, then is the following way correct to perform 100 training iteration on fed data:

```
# Run 100 x 10 iterations
for j in range(0, 100):
    data = np.random.rand(16, 100).astype(np.float32)
    label = (np.random.rand(16) * 10).astype(np.int32)

    workspace.FeedBlob("data", data)
    workspace.FeedBlob("label", label)

    workspace.RunNet(m.name, 10)   # run for 10 times
```
Where do we add createNet for this ?

Ques. 03
If above is true, I want to know, how do I save the trained model to run testing on my testing data ? Do we need to create separate NET for testing ?
Tutorials here do not explain it clearly : http://caffe2.ai/docs/intro-tutorial

If possible, can you give a complete example below with above requirements, Questions are mentioned in comments:

```
from caffe2.python import workspace, cnn
import numpy as np

#
#	Model Definition
#

# Create Model using a MODEL HELPER
model 	= cnn.CNNModelHelper(name="First Simple Model")

# Adding Fully Connected Layer to model object
fc_1  	= model.FC("data", "fc_1", dim_in=100, dim_out=10)	# Input Dimension = 100; Output Dimension = 10
								# Input = Data Blob; Output = FC_1 Blob

# Adding Sigmoid Layer to model
pred	= model.Sigmoid( fc_1, "pred")				# Input = FC_1 Blob; Output = pred Blob

# Adding Softmax with Loss Layer
[softmax, loss]	= model.SoftmaxWithLoss( [pred, "label"], ["softmax", "loss"])

# Initialize Parameters - One Time Step
workspace.RunNetOnce( model.param_init_net)

#
#       **Adding Entire Training Data to Workspace - For 100 iterations**
#
for j in range(0,100):
	# Create the input data for 
	data = np.random.rand(16, 100).astype(np.float32)

	# Create labels for the data as integers [0, 9].
	label = (np.random.rand(16) * 10).astype(np.int32)

	workspace.FeedBlob("data", data)
	workspace.FeedBlob("label", label)

#
#       Create Net
#
workspace.CreateNet(model.net)

#
#       Running Training Net - 100 iterations
#
for j in range(0,100):
	workspace.RunNet(model.name, 10)			# 10 times
        **#QUESTION                                               # Which Blob Does it Pick up everytime ?????????**

	print("\n Softmax:")
	print(workspace.FetchBlob("softmax"))

	print("\nLoss :")
	print(workspace.FetchBlob("loss"))


#
#       Testing Phase ?????? - What to do ????
#
```

**Comment by tarunjainsagar** [comment]
@Yangqing @bwasti Please Help. It is just a small example.

**Comment by cw-delli-bird** [comment]
I would like to know this as well. A matrix multiply using matmul would be extremely helpful.

**Comment by jiangnanhugo** [comment]
I also would like to know this. The paradigm of training seems to be blackbox, which is hard to set different strategies.

**Comment by lironmo** [comment]
same question!
more over where i can find the input paradigm? DatabaseOps  is not suitable to my case because i want to stream images and labels into the network, can i use tensor input instead the db op? i i set up the labels?
where i can find some good documentation the tutorial is shallow
